File: app/src/main/java/com/chiului/android_mvvm_architecture/ui/MainFragment.java
package com.chiului.android_mvvm_architecture.ui;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;

import com.chiului.android_mvvm_architecture.R;
import com.chiului.android_mvvm_architecture.adapter.MainAdapter;
import com.chiului.android_mvvm_architecture.base.BaseNavFragment;
import com.chiului.android_mvvm_architecture.databinding.FragmentMainBinding;
import com.chiului.android_mvvm_architecture.utilities.InjectorUtils;
import com.chiului.android_mvvm_architecture.utilities.ToastUtil;
import com.chiului.android_mvvm_architecture.viewmodel.MainViewModel;
import com.google.android.material.tabs.TabLayoutMediator;

/**
 * 主页碎片
 * @author 神经大条蕾弟
 * @date   2020/08/05 14:51
 */
public class MainFragment extends BaseNavFragment {

    private FragmentMainBinding mBinding;
    private MainViewModel mViewModel;

    @Override
    public int setContentViewID() {
        return R.layout.fragment_main;
    }

    @Override
    public View initViewModel(LayoutInflater inflater, int layoutId, @Nullable ViewGroup container) {
        mBinding = DataBindingUtil.inflate(inflater, layoutId, container, false);
        mBinding.setLifecycleOwner(this);
        mViewModel = new ViewModelProvider(this, InjectorUtils.INSTANCE.provideMainViewModelFactory(getActivity())).get(MainViewModel.class);

        mViewModel.getToast().observe(this, msg -> {
            ToastUtil.INSTANCE.show(getActivity(), msg, Toast.LENGTH_SHORT);
        });

        mViewModel.getUserInfoBean().observe(this, userBean -> {
            mViewModel.getToast().setValue(userBean.getNickName());
        });

        return mBinding.getRoot();
    }

    @Override
    public void initView() {
        setTopStack(true);
        initFragments();
        mViewModel.initUserInfo();
    }

    private void initFragments() {
        // 初始化适配器
        MainAdapter mainAdapter = new MainAdapter(this);

        // 给 ViewPager 设置适配器
        mBinding.viewPager.setAdapter(mainAdapter);

        // 把 TabLayout 与 ViewPager 绑定
        new TabLayoutMediator(mBinding.tabLayout, mBinding.viewPager, (tab, position) -> {
            tab.setIcon(getTabIcon(position));
            tab.setText(getTabText(position));
        }).attach();
    }

    /**
     * 根据 Position 获取 TabLayout 图标
     * @param position
     * @return
     */
    private int getTabIcon(int position){
        int resId = R.drawable.garden_tab_selector;
        switch (position){
            case 0:
                break;
            case 1:
                break;
            case 2:
                break;
            case 3:
                break;
            default:
        }
        return resId;
    }

    /**
     * 根据 Position 获取 TabLayout 文字
     * @param position
     * @return
     */
    private String getTabText(int position){
        String tab = "Module" + (position + 1);
        switch (position){
            case 0:
                break;
            case 1:
                break;
            case 2:
                break;
            case 3:
                break;
            default:
        }
        return tab;
    }

}

File: app/src/main/java/com/chiului/android_mvvm_architecture/ui/MainFragment.kt
package com.chiului.android_mvvm_architecture.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.viewModels
import com.chiului.android_mvvm_architecture.R
import com.chiului.android_mvvm_architecture.adapter.MainAdapter
import com.chiului.android_mvvm_architecture.base.BaseNavFragment
import com.chiului.android_mvvm_architecture.databinding.FragmentMainBinding
import com.chiului.android_mvvm_architecture.utilities.InjectorUtils
import com.chiului.android_mvvm_architecture.utilities.toast
import com.chiului.android_mvvm_architecture.viewmodel.MainViewModel
import com.google.android.material.tabs.TabLayoutMediator

/**
 * 主页碎片$
 * @author    神经大条蕾弟
 * @date      2021/02/03 10:15
 */
class MainFragment : BaseNavFragment() {

    lateinit var binding: FragmentMainBinding
    private val viewModel: MainViewModel by viewModels {
        InjectorUtils.provideMainViewModelFactory(requireActivity())
    }

    override fun initViewModel(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_main, container, false)
        binding.lifecycleOwner = this

        viewModel.toast.observe(this, {
            it.toast(requireContext())
        })

        viewModel.userInfoBean.observe(this, {
            viewModel.toast.value = it.nickName
        })

        return binding.root
    }

    override fun initView(view: View, savedInstanceState: Bundle?) {
        isTopStack = true
        initFragments()
        viewModel.initUserInfo()
    }

    private fun initFragments() {
        // 初始化适配器
        val mainAdapter = MainAdapter(this)

        // 给 ViewPager 设置适配器
        binding.viewPager.adapter = mainAdapter

        // 把 TabLayout 与 ViewPager 绑定
        TabLayoutMediator(binding.tabLayout, binding.viewPager) { tab, position ->
            tab.setIcon(getTabIcon(position))
            tab.setText(getTabText(position))
        }.attach()
    }

    /**
     * 根据 Position 获取 TabLayout 图标
     * @param position
     * @return
     */
    private fun getTabIcon(position: Int): Int {
        val resId = R.drawable.garden_tab_selector
        when (position) {
            0 -> {
            }
            1 -> {
            }
            2 -> {
            }
            3 -> {
            }
        }
        return resId
    }

    /**
     * 根据 Position 获取 TabLayout 文字
     * @param position
     * @return
     */
    private fun getTabText(position: Int): String {
        val tab = "Module" + (position + 1)
        when (position) {
            0 -> {
            }
            1 -> {
            }
            2 -> {
            }
            3 -> {
            }
        }
        return tab
    }

}